File: JavaScript/007-10001st-prime.js
/* 
	By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

	What is the nth prime number?
*/

function nthPrime(input) {

	// Set the first number to the first prime factor and add it to the list of primes
	let num = 2,
		primes = [2];

	// Keep going until we have an amount of primes equal to the input
	while(primes.length < input) {
		// Make a copy of the original number to manipulate it by dividing it by prime factors
		let currentNum = num;
		let i = 2;

		// Keep going until the number has been divided by all factors and is 1
		while(currentNum > 1) {

			// Check whether the number is divisible by the current "i" value, if yes, then "i" is a prime factor
			if(currentNum % i === 0) {

				currentNum /= i; // Update the current number by dividing it by the current factor "i"

				if(!primes.includes(i)) primes.push(i); // Add this prime factor to the list if it doesn't already exist
				// Don't change the counter, because the new number could also be divisible by it
			} else {
				i++; // Increment if the current number isn't divisible by the factor any more (or if it's not even a prime factor)
			}

		}

		num++;
	}
	// Return the last prime in the list, which will be the nth prime
	return primes.pop();

}
  
nthPrime(10001);
  

File: Ruby/007-10001st-prime.rb
# By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

# What is the 10 001st prime number?
# ---------------------------------------------------------------------------------------------------------
require 'prime'

def prime_at(prime_position)
  array_of_prime = []       # Maintain an array of prime numbers to keep track of length
  counter = 0               # Each index represents the position of the prime number
                            # The last element will be where we grab our prime number

  while(array_of_prime.length < prime_position)         # While we are not yet at the position we are looking for
    if (counter != 1) && (counter != prime_position)    # Keep finding the next prime  number
      if counter.prime?
        array_of_prime.push(counter)
      end
    end
    counter += 1
  end
  array_of_prime[-1]                                    # Return the final prime number
end

puts prime_at(10001)